### PR TITLE
Time-constrain user commands in XHarness workloads

### LINKED ISSUE #7899: Make sure XHarness telemetry is sent even when job times out
### Context
When we call user commands inside of the Helix SDK, we have to make sure that when they time out, we still send the telemetry at the end of the job (we do it via `HelixPostCommands`).

### To be done
- Investigate behavior of `HelixPostCommands` and whether they get called even when command times out
- Make sure telemetry is sent, i.e. time-constrain user commands and leave buffer to send the telemetry at the end

### PR DESCRIPTION
**Pull request by premun**
We have to make sure that user commands are time-constrained so that we have buffer to call clean-up and send the telemetry at the end of the job.

We do this by bumping the original work item timeout and setting a timeout for the injected user commands.

Resolves #7899

### CODE REVIEW

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1 [hunk 1]
```diff
@@ -35,11 +34,52 @@ function xharness() {
...
+do
+{
+    if ($task.IsCompleted) {
+        $exit_code = $code.ExitCode;
+        break;
+    }
 
-$exit_code=$LASTEXITCODE
+    if ($timer.Elapsed.TotalSeconds -gt $command_timeout) {
+        Write-Error "User command timed out after $command_timeout seconds!";
```
Comment (premun) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1:68]: @greenEkatherine later this is where we can add the clean-up (or we could just do it every time)

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1 [hunk 2]
```diff
@@ -35,11 +34,52 @@ function xharness() {
...
+}).AddArgument($PSScriptRoot).AddArgument($Env:HELIX_WORKITEM_UPLOAD_ROOT).AddArgument($app).AddArgument($timeout).AddArgument($command_timeout).AddArgument($package_name).AddArgument($expected_exit_code).AddArgument($device_output_path).AddArgument($instrumentation).AddArgument($code)
 
-$ErrorActionPreference="Continue"
+$output = New-Object 'System.Management.Automation.PSDataCollection[psobject]'
+$task = $job.BeginInvoke($output, $output);
+$timer = [Diagnostics.Stopwatch]::StartNew();
+
+do
+{
+    if ($task.IsCompleted) {
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1:62]: Seems OK but have you checked whether this works when the invocation writes to both StdOut & Err?  can have some unexpected behaviors that are missed when Err isn't written-to.

Comment (tkapin) [reply src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1:62]: This implementations with loop and sleep might work, but I'd recommend something like the following using proper Wait-Process call. IT should be even possible to get stderr and stdout properly.

```powershell
$timeout = 1
$psinfo = [System.Diagnostics.ProcessStartInfo]::new()
$psinfo.FileName = "ping.exe"
$psinfo.RedirectStandardError = $true
$psinfo.RedirectStandardOutput = $true
$psinfo.UseShellExecute = $false
$psinfo.Arguments = "localhost"
$p = [System.Diagnostics.Process]::new()
$p.StartInfo = $psinfo
$p.Start()
Wait-Process -InputObject $p -TimeOut $timeout -ErrorVariable ev -ErrorAction SilentlyContinue
if ($ev) {
        Stop-Process -InputObject $p -Force
        echo "Timeouted!"
}
$stdout = $p.StandardOutput.ReadToEnd()
$stderr = $p.StandardError.ReadToEnd()
Write-Host "stdout: $stdout"
Write-Host "stderr: $stderr"
Write-Host "exit code: " + $p.ExitCode
```

Comment (premun) [reply src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1]: Yeah, this sounds better. I also needed to make some other PowerShell methods available for the user, so I used this method + turned users' commands into a full cmdlet.

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1 [hunk 3]
```diff
@@ -22,24 +24,33 @@ param (
...
+$psinfo.FileName = "powershell"
+$psinfo.Arguments = " -ExecutionPolicy ByPass -NoProfile -File `"$PSScriptRoot\command.ps1`" -output_directory `"$Env:HELIX_WORKITEM_UPLOAD_ROOT`" -app `"$app`" -timeout `"$timeout`" -package_name `"$package_name`" -expected_exit_code `"$expected_exit_code`" -device_output_path `"$device_output_path`" -instrumentation `"$instrumentation`""
+$psinfo.RedirectStandardError = $false
+$psinfo.RedirectStandardOutput = $false
+$psinfo.UseShellExecute = $false
 
-$ErrorActionPreference="Continue"
+$process = [System.Diagnostics.Process]::new()
+$process.StartInfo = $psinfo
+$process.Start()
```
Comment (lpatalas) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1:40]: - Don't we lose all output from the process in this implementation?
- Also it seems that this whole block could be replaced by `Start-Process` command but that's just a cosmetic change

Comment (premun) [reply src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1:40]: - The output is there! I care about it very much!
- I was trying to find how to set some thing such as "UseShellExecute" through that and didn't find a way. I didn't find a way how to supply the StartInfo to it neither. Am I missing something?

Comment (lpatalas) [reply src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.ps1:40]: No, actually I was missing something in both cases :) I didn't know that by default output goes to the current console and you're right on the second point as well.

**Review by lpatalas — APPROVED** [review]

**Pull request merged**